tensorboard/plugins/distribution/polymer3/vz_distribution_chart/vz-distribution-chart.ts

```diff
@@ -12,234 +12,260 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-      return new Plottable.Components.Group(plots);
-    }
+  public setVisibleSeries(runs: string[]) {
+    this.runs = runs;
+    let datasets = runs.map((r) => this.getDataset(r));
+    this.plots.forEach((p) => p.datasets(datasets));
+  }
 
-    public setVisibleSeries(runs: string[]) {
-      this.runs = runs;
-      let datasets = runs.map((r) => this.getDataset(r));
-      this.plots.forEach((p) => p.datasets(datasets));
-    }
+  /**
+   * Set the data of a series on the chart.
+   */
+  public setSeriesData(name: string, data: any) {
+    this.getDataset(name).data(data);
+  }
+
+  public renderTo(targetSVG: d3.Selection<any, any, any, any>) {
+    this.targetSVG = targetSVG;
+    this.outer.renderTo(targetSVG);
+  }
 
-    /**
-     * Set the data of a series on the chart.
-     */
-    public setSeriesData(name: string, data: any) {
-      this.getDataset(name).data(data);
+  public redraw() {
+    this.outer.redraw();
+  }
+
+  protected destroy() {
+    this.outer.destroy();
+  }
+}
+
+export interface VzDistributionChart extends HTMLElement {
+  setSeriesData(name: string, data: unknown): void;
+  setVisibleSeries(names: readonly string[]): void;
+  redraw(): void;
+}
+
+@customElement('vz-distribution-chart')
+class _VzDistributionChart extends LegacyElementMixin(PolymerElement)
+  implements VzDistributionChart {
+  static readonly template = html`
+    <style include="plottable-style"></style>
+    <div id="chartdiv"></div>
+    <style>
+      :host {
+        -webkit-user-select: none;
+        -moz-user-select: none;
+        display: flex;
+        flex-direction: column;
+        flex-grow: 1;
+        flex-shrink: 1;
+        position: relative;
+      }
+      #chartdiv {
+        -webkit-user-select: none;
+        -moz-user-select: none;
+        flex-grow: 1;
+        flex-shrink: 1;
+      }
+    </style>
+  `;
+
+  @property({type: Object})
+  colorScale: Plottable.Scales.Color = new Plottable.Scales.Color().range(
+    d3.schemeCategory10.slice()
+  );
+
+  @property({type: String})
+  xType: string = 'step';
+
+  @property({type: Boolean})
+  _attached: boolean;
+
+  @property({type: Object})
+  _chart: DistributionChart;
+
+  @property({type: Array})
+  _visibleSeriesCache: string[] = [];
+
+  @property({type: Object})
+  _seriesDataCache: object = {};
+
+  @property({type: Number})
+  _makeChartAsyncCallbackId: number = null;
+
+  setVisibleSeries(names) {
+    this._visibleSeriesCache = names;
+    if (this._chart) {
+      this._chart.setVisibleSeries(names);
+      this.redraw();
     }
+  }
 
-    public renderTo(targetSVG: d3.Selection<any, any, any, any>) {
-      this.targetSVG = targetSVG;
-      this.outer.renderTo(targetSVG);
+  setSeriesData(name, data) {
+    this._seriesDataCache[name] = data;
+    if (this._chart) {
+      this._chart.setSeriesData(name, data);
     }
+  }
+
+  redraw() {
+    this._chart.redraw();
+  }
+
+  ready() {
+    super.ready();
+    this.scopeSubtree(this.$.chartdiv, true);
```

Comment:
Optional, but it looks like `scopeSubtree` is a no-op for Polymer 3, so this line (and the entire `ready()` override) could be omitted: https://polymer-library.polymer-project.org/3.0/api/legacy/legacy-element-mixin#LegacyElementMixin-method-scopeSubtree

I tested that locally and the dashboard still rendered normally.

Reply:
Okay, I guess that’s fine with me since it’s literally a no-op. Done.